File: solutions/java/0151.java
class Solution {
  public String reverseWords(String s) {
    char[] charArray = s.toCharArray();

    reverse(charArray, 0, charArray.length - 1);
    reverseWords(charArray, 0, 0);

    return cleanSpaces(charArray, 0, 0);
  }

  private void reverseWords(char[] charArray, int i, int j) {
    while (i < charArray.length) {
      while (i < j || i < charArray.length && charArray[i] == ' ')
        ++i;
      while (j < i || j < charArray.length && charArray[j] != ' ')
        ++j;
      reverse(charArray, i, j - 1);
    }
  }

  private String cleanSpaces(char[] charArray, int i, int j) {
    while (j < charArray.length) {
      while (j < charArray.length && charArray[j] == ' ')
        ++j;
      while (j < charArray.length && charArray[j] != ' ')
        charArray[i++] = charArray[j++];
      while (j < charArray.length && charArray[j] == ' ')
        ++j;
      if (j < charArray.length)
        charArray[i++] = ' ';
    }

    return new String(charArray).substring(0, i);
  }

  private void reverse(char[] charArray, int l, int r) {
    while (l < r) {
      char temp = charArray[l];
      charArray[l++] = charArray[r];
      charArray[r--] = temp;
    }
  }
}

File: solutions/cpp/0151.cpp
class Solution {
 public:
  string reverseWords(string s) {
    reverse(begin(s), end(s));
    reverseWords(s, 0, 0);

    return cleanSpaces(s, 0, 0);
  }

 private:
  void reverseWords(string& s, int i, int j) {
    while (i < s.length()) {
      while (i < j || i < s.length() && s[i] == ' ') ++i;
      while (j < i || j < s.length() && s[j] != ' ') ++j;
      reverse(begin(s) + i, begin(s) + j);
    }
  }

  string cleanSpaces(string& s, int i, int j) {
    while (j < s.length()) {
      while (j < s.length() && s[j] == ' ') ++j;
      while (j < s.length() && s[j] != ' ') s[i++] = s[j++];
      while (j < s.length() && s[j] == ' ') ++j;
      if (j < s.length()) s[i++] = ' ';
    }

    return s.substr(0, i);
  }
};